### DISCUSSION TITLE
Why WhenAnyValue stops working after using SuppressChangeNotifications?

### DISCUSSION THREAD

**Discussion by sinatrocious** [reactiveui/ReactiveUI#3830] [Q&A]
I have reactive object and monitoring for its reactive properties using `WhenAnyValue`. It works until I use `SuppressChangeNotifications` once. Then it become broken and I am trying to figure out what to do:

![reactive bug](https://github.com/reactiveui/ReactiveUI/assets/91904521/98783d05-5818-41b4-b905-e04b79a55b52)

<details>
  <summary>Show/hide repro</summary>

```c#
public class MainViewModel : ReactiveObject
{
    [Reactive]
    public string? Text { get; set; }

    [Reactive]
    public bool Value { get; set; }

    public ReactiveCommand<Unit, Unit> Test { get; }

    public MainViewModel()
    {
        this.WhenAnyValue(o => o.Value).Subscribe(value => Text = value.ToString());

        Test = ReactiveCommand.Create(() =>
        {
            using var dispose = this.SuppressChangeNotifications();
            Value = true;
        });
    }
}
```

```xaml
    <StackPanel>
        <CheckBox Content="{Binding Text}" IsChecked="{Binding Value}"/>
        <Button Content="Test" Command="{Binding Test}"/>
    </StackPanel>
```

</details>

After clicking `Test` button the checkbox is not updated for several clicks. How to fix this?

In real project there are many properties monitored in several places outside and I need to silently update them (load values without anyone noticing the changes) and I don't want to introduce `bool IgnoreChanges` and use it everywhere. This is where I was hoping `SuppressChangeNotifications` will work, but there is something like caching values and skipping emitting going on as it seems. Any easy workaround?

**Comment by anaisbetts** [comment]
I'm confused as to what you're finding unexpected here - you are suppressing change notifications, they shouldn't fire!

**Reply by sinatrocious** [reply]
I expect to see "True" text when checkbox is checked or "False" text when it's unchecked. That's not true for **two** clicks after suppression.

Put 3 breakpoints: one in value setter (V), one in text setter (T), one in subscribe (S).

1. Click checkbox - VST
2. Click checkbox - VST
3. Click `Test` button - V
4. Click checkbox - V
5. Click checkbox - V
6. Click checkbox - VST

Why 4 and 5 step are so strange?

**Comment by ChrisPulman** [comment]
You are waiting for Garbage Collection to remove the Suppression after clicking the button, as your not disposing immediately the suppression will remain active until the Garbage Collector comes along to clean up, hence the delay.

**Reply by sinatrocious** [reply]
The dispose method is called at the end of scope, which in this case is end of lambda. [Worth to read](https://stackoverflow.com/q/45036/1997232) about GC.

**Reply by ChrisPulman** [reply]
And if you use:
using (var dispose = this.SuppressChangeNotifications())
{
            Value = true;
}

Is it the same result?

**Reply by sinatrocious** [reply]
Yes, same. VS prompts me with [IDE0063](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0063).

**Reply by ChrisPulman** [reply]
Okay, I will take a look at this later when I finish work.

**Reply by ChrisPulman** [reply]
Okay, the binding is a bool, WhenAnyValue gives change notifications so therefore this is the way it currently operates:

using this code

```
public MainViewModel()
{
    var count = 0;
    this.WhenAnyValue(o => o.Value).Subscribe(value => Text = value.ToString() + count++);

    Test = ReactiveCommand.Create(() =>
    {
        using var dispose = SuppressChangeNotifications();
        Value = true;
        count = 0;
    });
}
```

Scenario 1 - Suppression When FALSE with a suppressed TRUE change
```
Unchecked => changes false
Checked => changes true
Unchecked => changes false
SuppressChangeNotifications
{
Set Value true => no change as suppressed
Dispose Suppression
}
Checked => no change due to equal to suppressed value
Unchecked => no change as equal to last change
Checked => changes true
```

Scenario 2 - Suppression When TRUE with a suppressed TRUE change

```
Unchecked => changes false
Checked => changes true
SuppressChangeNotifications
{
Set Value true => no change as suppressed
Dispose Suppression
}
Unchecked => changes false as not equal to last change
Checked => changes true
```

using this code

```
public MainViewModel()
{
    var count = 0;
    this.WhenAnyValue(o => o.Value).Subscribe(value => Text = value.ToString() + count++);

    Test = ReactiveCommand.Create(() =>
    {
        using var dispose = SuppressChangeNotifications();
        Value = false;
        count = 0;
    });
}
```

Scenario 3 - Suppression When FALSE with a suppressed FALSE change
```
Unchecked => changes false
Checked => changes true
Unchecked => changes false
SuppressChangeNotifications
{
Set Value false=> no change as suppressed
Dispose Suppression
}
Checked => changes true
Unchecked => changes false
```

Scenario 4 - Suppression When TRUE with a suppressed FALSE change

```
Unchecked => changes false
Checked => changes true
SuppressChangeNotifications
{
Set Value false=> no change as suppressed
Dispose Suppression
}
Unchecked => no change due to equal to suppressed value
Checked => no change as equal to last change
Unchecked => changes false
```

**Reply by sinatrocious** [reply]
Hmm.. "no change as equal to last change" happens because the change while suppressed will cause same value to be ignored and not remembered as "last change". That make sense, thanks.

Is it possible to update that "last changed" value while suppressing? My intent is to update value without triggering any `WhenAnyValue().Subscribe()`, but any change after suppression should work without problems! I am willing to add some logic where I suppress change, but I'd like to keep subscribers code clean.

As I said earlier, I don't want to introduce `bool` flag to be checked by every subscriber, whenever they need to ignore change made while flag is set:

<details>
  <summary>Show/hide code with bool flag</summary>

```c#
        this.WhenAnyValue(o => o.Value).Subscribe(value =>
        {
            if(!IgnoreValue) // suppressed changes are ignored by subscriber
                Text = value.ToString();
        });

        Test = ReactiveCommand.Create(() =>
        {
            IgnoreValue = true; // suppress changes by setting flag
            Value = true;
            IgnoreValue = false;
        });
```
</details

**Reply by sinatrocious** [reply]
I've refactored the code with bool flag and it looks not so bad:

<details>
  <summary>Show/hide refactored code</summary>

```c#
public class MainViewModel : ReactiveObject, ISuspendable
{
    [Reactive]
    public string? Text { get; set; }

    [Reactive]
    public bool Value { get; set; }

    public ReactiveCommand<Unit, Unit> Test { get; }

    bool ISuspendable.IsSuspended { get; set; }

    public MainViewModel()
    {
        this.WhenAnyValue(o => o.Value).Subscribe(this, value => Text = value.ToString());

        Test = ReactiveCommand.Create(() =>
        {
            using var dispose = this.Suspend();
            Value = true;
        });
    }
}

public interface ISuspendable
{
    bool IsSuspended { get; set; }
}

public sealed class SuspendableDisposeHelper : IDisposable
{
    readonly ISuspendable suspendable;

    public SuspendableDisposeHelper (ISuspendable suspendable)
    {
        this.suspendable = suspendable;
        suspendable.IsSuspended = true;
    }

    public void Dispose() => suspendable.IsSuspended = false;
}

public static class SuspendableExtensions
{
    public static IDisposable Suspend(this ISuspendable suspendable) => new SuspendableDisposeHelper(suspendable);

    public static IDisposable Subscribe<T>(this IObservable<T> source, ISuspendable suspendable, Action<T> onNext) =>
        source.Subscribe(o =>
        {
            if (!suspendable.IsSuspended)
                onNext(o);
        });
}
```
</details>

There is one more parameter to pass into `Subscribe` method, but otherwise the code looks similar to one with `SuppressChangeNotifications`.
